struct.go

```diff
@@ -81,7 +117,12 @@ func ValidateStruct(structPtr interface{}, fields ...*FieldRules) error {
 		if ft == nil {
 			return NewInternalError(ErrFieldNotFound(i))
 		}
-		if err := Validate(fv.Elem().Interface(), fr.rules...); err != nil {
+
+		// a field can have rules OR context rules.
+		if fr.rulesWithContext != nil && fr.rules != nil {
```

Comment:
Why a field cannot have both kinds of rules?

Reply:
If you have more than one list of rules for a field, features like Skip don't fit very well, as the ordering of rules for the field is now split across two lists by type.

Reply:
How about keeping both kinds of rules in a single list (and thus maintaining their declaration order)?
* When ValidateStructWithContext() is called, if a rule implements RuleWithContext, it will be called with the context. Otherwise, it will be called without the context.
* When ValidateStruct() is called, rules that only implement RuleWithContext will be called with a background context.

Reply:
By doing so, rules like `required` no longer needs to implement `RuleWithContext`.

Reply:
That sounds great, but does that not mean losing type safety of the rules slice in FieldRules?...  `FieldRules struct {
		fieldPtr         interface{}
		rules            []interface{}
	}`

Reply:
I think that's fine because this is private, and the functions that modify this field ensure it only holds expected types.

Reply:
However, I find that this affects the ValidateWithContext and Validate function signatures, which are not private.  For example, you would need... `func ValidateWithContext(ctx context.Context, value interface{}, rules ...**interface{}**) error {...`

Reply:
Hmm. I see it only affects calls to these interfaces from internal code, so i could use reflection to work out what is in the slice and convert to the correctly typed slice before calling ValidateWithContext / Validate.  However, I think this potentially a bit murky vs the current method.  Currently, I believe it is clear that you setup the FieldRules with either FieldWithContext or Field.  And it is easy to tell which method was used by checking which slice  is not nil.

Reply:
@qiangxue are you able to illustrate how this could be done without compromising any external interfaces? it's not clear to me how this can be done.